packages/plugin-core/src/commands/GotoNote.ts

```diff
@@ -78,9 +88,97 @@ export class GotoNoteCommand extends BasicCommand<CommandOpts, CommandOutput> {
     };
   }
 
+  private getQs(opts: CommandOpts, link: FoundLinkSelection): CommandOpts {
+    if (link.value) {
+      // Reference to another file
+      opts.qs = link.value;
+    } else {
+      // Same file block reference, implicitly current file
+      const note = WSUtils.getActiveNote();
+      if (note) {
+        // Same file link within note
+        opts.qs = note.fname;
+        opts.vault = note.vault;
+      } else {
+        const { wsRoot, vaults } = getDWorkspace().engine;
+        // Same file link within non-note file
+        opts.qs = path.relative(
+          wsRoot,
+          VSCodeUtils.getActiveTextEditorOrThrow().document.fileName
+        );
+        opts.vault = VaultUtils.getVaultByNotePath({
+          wsRoot,
+          vaults,
+          fsPath: opts.qs,
+        });
+      }
+    }
+    return opts;
+  }
+
+  private async getExistingNote(opts: CommandOpts) {
+    const { engine } = getDWorkspace();
+    const notes = NoteUtils.getNotesByFname({
+      fname: opts.qs!,
+      notes: engine.notes,
+    });
+    if (notes.length === 1) {
+      // There's just one note, so that's the one we'll go with.
+      opts.vault = notes[0].vault;
+    } else if (notes.length > 1) {
+      // It's ambiguous which note the user wants to go to, so we have to
+      // guess or prompt.
+      const resp = await PickerUtilsV2.promptVault(
+        notes.map((ent) => ent.vault)
+      );
+      if (_.isUndefined(resp)) return null;
```

Comment:
since the normal pattern for cancelled input is `undefined` for quickpick related events, why not just pass through `undefined` here?

Reply:
If the user cancelled the prompt, then we want to stop and not try the other options. So we have to differentiate between "couldn't find a note" and "found multiple notes, but user cancelled the prompt". I thought returning null might be a good way to signal that, it's not just unset (undefined) but explicitly refused.